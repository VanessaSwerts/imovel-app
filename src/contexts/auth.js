import React, { createContext, useContext, useState, useEffect } from 'react'
import api from '../services/api'
import AsyncStorage from '@react-native-community/async-storage'
import * as SplashScreen from 'expo-splash-screen'

const AuthContext = createContext({
  signed: false,
  user: {}
})

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null)

  useEffect(() => {
    async function loadStorageData() {
      await SplashScreen.preventAutoHideAsync()

      const [token, userData] = await AsyncStorage.multiGet([
        'user-token',
        'user-data'
      ])

      if (token[1] && userData[1]) {
        api.defaults.headers.Authorization = `Bearer ${token[1]}`
        setUser(JSON.parse(userData[1]))
      }

      await SplashScreen.hideAsync()
    }

    loadStorageData()
  }, [])

  const signIn = async (email, password) => {
    return new Promise((resolve, reject) => {
      const data = {
        email,
        password
      }

      api
        .post('/auth/login', data)
        .then(async res => {
          const { token, ...userData } = res.data

          api.defaults.headers.Authorization = `Bearer ${token}`

          setUser(userData)

          resolve(res)

          await AsyncStorage.setItem('user-token', token)
          await AsyncStorage.setItem('user-data', JSON.stringify(userData))
        })
        .catch(err => {
          reject(err)
          console.error(err)
        })
    })
  }

  const signOut = async () => {
    await AsyncStorage.clear()

    api.defaults.headers.Authorization = ''

    setUser(null)
  }

  return (
    <AuthContext.Provider value={{ signed: !!user, user, signIn, signOut }}>
      {children}
    </AuthContext.Provider>
  )
}

export function useAuth() {
  return useContext(AuthContext)
}
