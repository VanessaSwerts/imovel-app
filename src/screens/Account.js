import React, { useState, useEffect } from 'react'
import {
  SafeAreaView,
  View,
  Dimensions,
  StyleSheet,
  TouchableOpacity,
  Text,
  Image,
  KeyboardAvoidingView,
  Platform
} from 'react-native'
import { useNavigation } from '@react-navigation/native'
import api, { STORAGE_URL } from '../services/api'
import AsyncStorage from '@react-native-community/async-storage'

import colors from '../constants/colors.json'

import EditTextField from '../components/EditTextField'

const CARD_HEIGHT = Dimensions.get('window').height * 0.34

export default function Account() {
  const navigation = useNavigation()

  const [userInfo, setUserInfo] = useState({
    name: "",
    email: "",
    phone: "",
    avatar: "default-avatar.png"
  })

  const getUser = async () => {
    const token = await AsyncStorage.getItem("user-token")
    if (!token) {
      navigation.navigate("SignIn", { backPath: "Minha conta" })
      return
    }

    const config = {
      headers: {
        "Authorization": "Bearer " + token
      }
    }

    api.get(`/user`, config)
      .then((res) => {
        setUserInfo(res.data)
      })
      .catch((err) => {
        console.error(err)
      })
  }

  const logout = async () => {
    await AsyncStorage.clear()
    navigation.navigate("SignIn", { backPath: "Minha conta" })
  }

  useEffect(() => {
    getUser()
  }, [])

  return (
    <View style={styles.container}>
      <SafeAreaView style={styles.header}>
        <Text style={styles.title}> Minha conta </Text>

        <View style={styles.cardContainer}>
          <Image style={styles.avatarContainer} resizeMode="contain"
            source={{ uri: `${STORAGE_URL}/user/${userInfo.avatar}` }} />
          <Text style={styles.name}>{userInfo.name}</Text>
          <TouchableOpacity style={styles.buttonLogout} onPress={() => logout()}>
            <Text style={styles.buttonText}>Sair</Text>
          </TouchableOpacity>
        </View>
      </SafeAreaView>

      <KeyboardAvoidingView>
        <KeyboardAvoidingView style={styles.body}
          behavior="padding"
          enabled={Platform.OS === 'ios'}
        >
          <View>
            <EditTextField label={'Nome: '} text={'Pedro Henrique Silva'} edit={true} />
            <EditTextField label={'E-mail: '} text={'pedrohs@gmail.com'} keyboardType={'email-address'} edit={true} />
            <EditTextField label={'Celular: '} text={'(35) 99988-7766'} keyboardType={'phone-pad'} edit={true} />
            <TouchableOpacity style={styles.button}>
              <Text style={styles.buttonText}>Salvar</Text>
            </TouchableOpacity>
          </View>
        </KeyboardAvoidingView>
      </KeyboardAvoidingView>
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors["platinum"],
  },

  header: {
    paddingVertical: 10,
    backgroundColor: colors["blue-secondary"],
  },

  title: {
    margin: 10,
    fontSize: 28,
    fontWeight: "500",
    color: colors['yellow'],
    alignSelf: 'center',
  },

  cardContainer: {
    width: 330,
    alignSelf: 'center',
    backgroundColor: "#FFF",
    borderWidth: 1,
    borderRadius: 24,
    borderColor: '#b2b2b2',
    overflow: 'hidden',
    marginTop: 10,
    padding: 10,
    flexDirection: 'column',
    alignItems: 'center',
  },

  avatarContainer: {
    width: 115,
    height: 115,
    borderRadius: 100,
    borderWidth: 3,
    borderColor: colors["blue"],
    backgroundColor: colors["blue"],
  },

  name: {
    marginTop: 10,
    fontSize: 18,
    fontWeight: "500",
    color: colors['blue'],
    alignSelf: 'center',
  },

  city: {
    marginTop: 5,
    fontSize: 12,
    fontWeight: "300",
    color: colors['blue'],
    alignSelf: 'center',
  },

  buttonLogout: {
    height: 30,
    width: 100,
    backgroundColor: colors['yellow'],
    borderRadius: 24,
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
    marginVertical: 10
  },

  editIcon: {
    position: 'absolute',
    right: 18,
    top: 18,
  },

  body: {
    height: '100%',
    paddingHorizontal: 20,
    marginTop: CARD_HEIGHT * 0.3,
  },

  button: {
    height: 35,
    width: 150,
    maxWidth: 200,
    backgroundColor: colors['yellow'],
    borderRadius: 24,
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
    marginTop: 30,
  },

  buttonText: {
    color: colors['blue'],
    fontWeight: 'bold',
    fontSize: 16,
  },

})