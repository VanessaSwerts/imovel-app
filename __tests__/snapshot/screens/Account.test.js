import React from 'react'
import * as auth from '../../../src/contexts/auth'
import { create } from 'react-test-renderer'
import Account from '../../../src/screens/Account'

const mockedNavigate = jest.fn()

jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: () => ({
      navigate: mockedNavigate
    }),
    useRoute: () => ({
      route: jest.fn()
    })
  }
})

describe('Account unit test', () => {
  it('render Account screen correctly when signed=false', () => {
    jest.spyOn(auth, 'useAuth').mockImplementation(() => ({
      signed: false
    }))
    const tree = create(<Account />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
  it('render Account screen correctly when signed=true', async () => {
    jest.spyOn(auth, 'useAuth').mockImplementation(() => ({
      signed: true
    }))

    const tree = create(<Account />)
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
