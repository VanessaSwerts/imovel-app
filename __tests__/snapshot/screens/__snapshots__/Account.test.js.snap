// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account unit test render Account screen correctly when signed=false 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-between",
      "paddingHorizontal": "10%",
      "paddingVertical": "25%",
    }
  }
  testID="empty-message"
>
  <Icon
    color="#14213D"
    name="user-alt"
    size={120}
  />
  <View>
    <Text
      style={
        Object {
          "color": "#333740",
          "fontSize": 20,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Você ainda não está logado em uma conta!
    </Text>
    <Text
      style={
        Object {
          "color": "#AAADB3",
          "fontSize": 18,
          "marginTop": 15,
          "textAlign": "center",
        }
      }
    >
      Faça o login no aplicativo para poder acessar os dados da sua conta.
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#19294d",
        "borderRadius": 24,
        "height": 40,
        "justifyContent": "center",
        "marginVertical": 10,
        "opacity": 1,
        "paddingHorizontal": 25,
        "width": 130,
      }
    }
    testID="account-empty-button"
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
        }
      }
    >
      Entrar
    </Text>
  </View>
</View>
`;

exports[`Account unit test render Account screen correctly when signed=true 1`] = `
<RCTScrollView
  alwaysBounceVertical={false}
  showsVerticalScrollIndicator={false}
  style={
    Object {
      "backgroundColor": "#E5E5E5",
      "flex": 1,
    }
  }
  testID="account"
>
  <View>
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "#E5E5E5",
            "flex": 1,
          },
          Object {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#14213D",
            "height": "40%",
            "position": "absolute",
            "width": "100%",
          }
        }
      />
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#FCA311",
            "fontSize": 28,
            "fontWeight": "600",
            "margin": 10,
            "paddingTop": 30,
          }
        }
      >
        Minha conta
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#FFF",
            "borderColor": "#b2b2b2",
            "borderRadius": 24,
            "borderWidth": 1,
            "flexDirection": "column",
            "marginTop": 10,
            "overflow": "hidden",
            "padding": 8,
            "width": 330,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              Object {
                "uri": "https://res.cloudinary.com/imovel/image/upload/v1602331961/user/default-avatar.png",
              }
            }
            style={
              Object {
                "borderColor": "#FCA311",
                "borderRadius": 110,
                "borderWidth": 2,
                "height": 110,
                "width": 110,
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "alignSelf": "center",
              "color": "#14213D",
              "fontSize": 18,
              "fontWeight": "500",
              "marginTop": 5,
            }
          }
          testID="account-name"
        >
          
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#FCA311",
              "borderRadius": 24,
              "height": 30,
              "justifyContent": "center",
              "marginVertical": 10,
              "opacity": 1,
              "width": 100,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#14213D",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Sair
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "paddingHorizontal": 20,
            "paddingVertical": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "red",
              "fontSize": 14,
              "fontWeight": "500",
              "paddingTop": 10,
              "textAlign": "center",
            }
          }
          testID="account-errorMessageText"
        >
          
        </Text>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "backgroundColor": "#FFF",
                "borderColor": "#14213D",
                "borderRadius": 25,
                "borderWidth": 1,
                "elevation": 2,
                "flexDirection": "row",
                "marginTop": 15,
                "paddingHorizontal": 12,
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0.2,
                "shadowRadius": 2,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "#14213D",
                "fontWeight": "600",
              }
            }
          >
            Nome: 
          </Text>
          <TextInput
            allowFontScaling={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Seu nome..."
            placeholderTextColor="#d2d2d2"
            rejectResponderTermination={true}
            secureTextEntry={false}
            style={
              Array [
                Object {
                  "height": 46,
                  "paddingHorizontal": 8,
                  "paddingVertical": 10,
                  "width": "100%",
                },
                undefined,
              ]
            }
            testID="account-input-name"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "backgroundColor": "#FFF",
                "borderColor": "#14213D",
                "borderRadius": 25,
                "borderWidth": 1,
                "elevation": 2,
                "flexDirection": "row",
                "marginTop": 15,
                "paddingHorizontal": 12,
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0.2,
                "shadowRadius": 2,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "#14213D",
                "fontWeight": "600",
              }
            }
          >
            E-mail: 
          </Text>
          <TextInput
            allowFontScaling={true}
            keyboardType="email-address"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Seu email..."
            placeholderTextColor="#d2d2d2"
            rejectResponderTermination={true}
            secureTextEntry={false}
            style={
              Array [
                Object {
                  "height": 46,
                  "paddingHorizontal": 8,
                  "paddingVertical": 10,
                  "width": "100%",
                },
                undefined,
              ]
            }
            testID="account-input-email"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "backgroundColor": "#FFF",
                "borderColor": "#14213D",
                "borderRadius": 25,
                "borderWidth": 1,
                "elevation": 2,
                "flexDirection": "row",
                "marginTop": 15,
                "paddingHorizontal": 12,
                "shadowOffset": Object {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0.2,
                "shadowRadius": 2,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "color": "#14213D",
                "fontWeight": "600",
              }
            }
          >
            Celular: 
          </Text>
          <TextInput
            allowFontScaling={true}
            keyboardType="phone-pad"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Seu celular..."
            placeholderTextColor="#d2d2d2"
            rejectResponderTermination={true}
            secureTextEntry={false}
            style={
              Array [
                Object {
                  "height": 46,
                  "paddingHorizontal": 8,
                  "paddingVertical": 10,
                  "width": "100%",
                },
                undefined,
              ]
            }
            testID="account-input-phone"
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "#FCA311",
              "borderRadius": 24,
              "height": 35,
              "justifyContent": "center",
              "marginTop": 30,
              "maxWidth": 200,
              "opacity": 1,
              "width": 150,
            }
          }
          testID="account-save-button"
        >
          <Text
            style={
              Object {
                "color": "#14213D",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Salvar
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
